File: 2015/7/solution.java

import java.util.HashMap;
import java.util.Scanner;

public class AoC_07 {

    enum Logic {

        AND, OR, NOT, RSHIFT, LSHIFT, DIRECT;

        final static int WIDTH = 0xFFFF;

        int output(int[] input) {
            switch (this) {
                case AND:
                    return input[0] & input[1];
                case OR:
                    return input[0] | input[1];
                case NOT:
                    return (~input[0]) & WIDTH;
                case LSHIFT:
                    return (input[0] << input[1]) & WIDTH;
                case RSHIFT:
                    return input[0] >> input[1];
                default:
                    return input[0];
            }
        }
    };

    class Circuit {

        HashMap<String, Gate> gates = new HashMap<>();

        Gate getGate(String wireName) {
            if (!gates.containsKey(wireName)) {
                gates.put(wireName, new Gate());
            }
            return gates.get(wireName);
        }

        class Gate {

            Logic logic = null;
            Gate[] inputs;
            int[] inputValues;

            int getOutput() {
                for (int i = 0; i < inputValues.length; i++) {
                    if (inputValues[i] == -1 && inputs[i] != null) {
                        inputValues[i] = inputs[i].getOutput();
                    }
                }
                return logic.output(inputValues);
            }

            void setInput(String[] data) {
                inputs = new Gate[data.length];
                inputValues = new int[data.length];
                for (int i = 0; i < data.length; i++) {
                    try {
                        inputValues[i] = Integer.parseInt(data[i]);
                    } catch (Exception e) {
                        inputs[i] = getGate(data[i]);
                        inputValues[i] = -1;
                    }
                }
            }

            void reset() {
                for (int i = 0; i < inputValues.length; i++) {
                    inputValues[i] = inputs[i] != null ? -1 : inputValues[i];
                }
            }
        }
    }

    void magic() {
        Circuit circuit = new Circuit();
        while (sc.hasNextLine()) {
            String[] input = sc.nextLine().split(" ");
            Circuit.Gate gate = circuit.getGate(input[input.length - 1]);
            switch (input.length) {
                case 3:
                    gate.logic = Logic.DIRECT;
                    gate.setInput(new String[]{input[0]});
                    break;
                case 4:
                    gate.logic = Logic.NOT;
                    gate.setInput(new String[]{input[1]});
                    break;
                default:
                    gate.logic = Logic.valueOf(input[1]);
                    gate.setInput(new String[]{input[0], input[2]});
                    break;
            }
        }
        Circuit.Gate a = circuit.getGate("a");
        System.out.println("Part One: " + a.getOutput());

        Circuit.Gate b = circuit.getGate("b");
        b.inputs = new Circuit.Gate[1];
        b.inputValues = new int[]{a.getOutput()};
        b.logic = Logic.DIRECT;
        for (Circuit.Gate gate : circuit.gates.values()) {
            gate.reset();
        }
        System.out.println("Part Two: " + a.getOutput());
    }

    static Scanner sc = new Scanner(System.in);

    public static void main(String[] args) {
        long startTime = System.currentTimeMillis();
        new AoC_07().magic();
        System.out.println("Running time: " + (System.currentTimeMillis() - startTime) + "ms");
    }
}

File: 2015/7/solution.php
<?php

class Wire
{
    private $inputs;
    private $logic;
    private $output;

    public function __construct(string $logic, string ...$inputs)
    {
        $this->logic  = $logic;
        $this->inputs = $inputs;
    }

    private function value(string $input)
    {
        if (preg_match('/^\d+$/', $input)) {
            return intval($input);
        }

        global $wires;

        return $wires[$input]->output();
    }

    private function calculateOutput(): void
    {
        switch ($this->logic) {
            case 'DIRECT':
                $this->output = $this->value($this->inputs[0]);
                break;
            case 'NOT':
                $this->output = ~$this->value($this->inputs[0]);
                break;
            case 'AND':
                $this->output = $this->value($this->inputs[0]) & $this->value($this->inputs[1]);
                break;
            case 'OR':
                $this->output = $this->value($this->inputs[0]) | $this->value($this->inputs[1]);
                break;
            case 'LSHIFT':
                $this->output = $this->value($this->inputs[0]) << $this->value($this->inputs[1]);
                break;
            case 'RSHIFT':
                $this->output = $this->value($this->inputs[0]) >> $this->value($this->inputs[1]);
                break;
        }

        $this->output &= 0xFFFF;
    }

    public function output(): ?int
    {
        if (is_null($this->output)) {
            $this->calculateOutput();
        }

        return $this->output;
    }

    public function reset(): void
    {
        $this->output = null;
    }
}

$wires = [];

while ($line = trim(fgets(STDIN))) {
    $explode = explode(' ', $line);
    switch (count($explode)) {
        case 3:
            $wires[$explode[2]] = new Wire('DIRECT', $explode[0]);
            break;
        case 4:
            $wires[$explode[3]] = new Wire($explode[0], $explode[1]);
            break;
        case 5:
            $wires[$explode[4]] = new Wire($explode[1], $explode[0], $explode[2]);
            break;
    }
}

fwrite(STDOUT, $a = $wires["a"]->output() . PHP_EOL);

foreach ($wires as $wire) {
    $wire->reset();
}
$wires["b"] = new Wire('DIRECT', $a);

fwrite(STDOUT, $wires["a"]->output() . PHP_EOL);